File: backend/voting-app/src/main/java/com/tu/votingapp/utils/mappers/UserMapper.java
package com.tu.votingapp.utils.mappers;

import com.tu.votingapp.dto.general.UserDTO;
import com.tu.votingapp.entities.LocationEntity;
import com.tu.votingapp.entities.UserEntity;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;


@Mapper(componentModel = "spring", uses = {LocationMapper.class , DocumentMapper.class})
public interface UserMapper {

    // When mapping FROM Entity TO DTO:
    @Mapping(source = "regionId", target = "locationId") // Existing location mapping
    @Mapping(target = "document", ignore = true)      // <<<--- ADD THIS LINE TO IGNORE DOCUMENT
    @Mapping(target = "password", ignore = true)      // <<<--- Also good practice: ignore password
    UserDTO toDto(UserEntity user);

    // When mapping FROM DTO TO Entity:
    // Let MapStruct handle location. Document mapping might be handled manually
    // in the service layer when creating/updating if complex logic is needed,
    // or via another @Mapping if simple. Ignoring password is correct here too.
    @Mapping(source = "locationId", target = "regionId")
    @Mapping(target = "document", ignore = true)      // Often ignored here too, set manually in service
    @Mapping(target = "password", ignore = true)      // Password should be handled/hashed separately
    @Mapping(target = "roles", ignore = true)         // Roles are usually managed via specific service methods
    UserEntity toEntity(UserDTO userDto);
}

File: frontend/android/app/src/main/java/com/example/android/utils/mappers/UserMapper.kt
package com.example.android.utils.mappers

import com.example.android.dto.response.UserDTO
import com.example.android.entity.DocumentEntity
import com.example.android.entity.LocationEntity
import com.example.android.entity.MunicipalityEntity
import com.example.android.entity.RegionEntity
import com.example.android.entity.RoleEntity
import com.example.android.entity.UserEntity
import com.example.android.utils.mappers.DocumentMapper.toDto
import com.example.android.utils.mappers.LocationMapper.toDto
import com.example.android.utils.mappers.RoleMapper.toDto

object UserMapper {

    fun UserDTO.toEntity(): UserEntity = UserEntity(
        id = this.id?:1,
        name = this.name?:"",
        email = this.email?:"",
        phone = this.phone?:"",
        password = this.password?:"",
        currentAddress = this.currentAddress?:"",
        locationId = this.locationId?.id ?: 1L,
        egn = this.egn?:""
    )

    fun UserEntity.toDto(
        location: LocationEntity,
        municipalityEntity: MunicipalityEntity,
        regionEntity: RegionEntity,
        document: DocumentEntity?,
        roles: List<RoleEntity>
    ): UserDTO = UserDTO(
        id = this.id,
        name = this.name,
        email = this.email,
        phone = this.phone,
        password = this.password,
        currentAddress = this.currentAddress,
        egn = this.egn,
        locationId = location.toDto(
            municipality = municipalityEntity,
            region = regionEntity
        ),
        document = document?.toDto(),

        roles = roles[0].toDto()
    )
}